physx/src/scene.rs

```diff
@@ -10,487 +10,972 @@
+                .collect::<Vec<_>>()
+                .as_ptr(),
+            actors.len() as u32,
+            wake_touching,
+        );
+    }
+
+    /// Add an articulation to the world.
+    pub fn add_articulation(&mut self, mut articulation: Owner<impl ArticulationBase>) {
+        unsafe {
+            PxScene_addArticulation_mut(self.as_mut_ptr(), articulation.as_mut_ptr());
+            forget(articulation);
+        };
     }
 
     /// Remove an articulation from the world
-    pub fn remove_articulation(&mut self, handle: BodyHandle) {
+    pub fn remove_articulation(
+        &mut self,
+        articulation: &mut impl ArticulationBase,
+        wake_touching: bool,
+    ) {
         unsafe {
-            let articulation = handle.0 as *const PxArticulationBase;
             PxScene_removeArticulation_mut(
-                self.px_scene.write().unwrap().expect("accessing null ptr"),
-                articulation as *mut PxArticulationBase,
-                false,
+                self.as_mut_ptr(),
+                articulation.as_mut_ptr(),
+                wake_touching,
             );
+        }
+    }
 
-            if let Some(idx) = self
-                .bodies
-                .iter()
-                .position(|b| b.deref().get_raw() == articulation)
-            {
-                let mut body = self.bodies.swap_remove(idx);
-                body.release();
+    pub fn add_aggregate(&mut self, mut aggregate: Owner<Aggregate<L, S, D, H, M>>) {
+        unsafe {
+            PxScene_addAggregate_mut(self.as_mut_ptr(), aggregate.as_mut_ptr());
+            forget(aggregate);
+        }
+    }
+
+    pub fn remove_aggregate(
+        &mut self,
+        aggregate: &mut Aggregate<L, S, D, H, M>,
+        wake_touching: bool,
+    ) {
+        unsafe {
+            PxScene_removeAggregate_mut(self.as_mut_ptr(), aggregate.as_mut_ptr(), wake_touching);
+        }
+    }
+
+    /* TODO implement Collection and the rest of the serialization pipeline
+        fn add_collection(&mut self, collection: &PxCollection) {
+            unsafe {
+                PxScene_addCollection_mut(
+                    self.as_mut_ptr(),
+                    collection
+                );
             }
         }
+    */
+
+    // TODO implement PruningStructure trait
+    pub fn add_pruning_structure(&mut self, mut pruning_structure: Owner<PruningStructure>) {
+        unsafe {
+            PxScene_addActors_mut_1(self.as_mut_ptr(), pruning_structure.as_mut_ptr());
+        }
+        forget(pruning_structure);
     }
 
     ////////////////////////////////////////////////////////////////////////////////
+    // Simulation
+    // TODO write a proper type for the scratch buffer.
+    pub fn simulate(
+        &mut self,
+        time_step: f32,
+        completion_task: Option<&mut PxBaseTask>,
+        // scratch memory must be 16-align, 16K blocks
+        scratch: Option<&mut [[(u64, u64); 1024]]>,
```

Comment:
This API is bad, a pair of u64 is align(8) 🤦.
I replaced it with an actual type that maintains the alignment and size invariants.  It's constructor is unsafe though, because the block needs to be valid until Scene::fetch_results is called.  I put the unsafe bound on the ctor instead of on Scene::simulate so the core API could be safe.